### PR TITLE
Fix for SCM-714: mvn release:prepare fails if the command line is too long on windows

### PR DESCRIPTION
**Pull request by murinrad**
This is a fix for SCM-714

This fix changes the behaviour of git add on Windows machines.
The git add is to be executed per file.

Fix tested on local and correctly splits the add command per file.

### CODE REVIEW

**Comment by murinrad** [comment]
Ticket in the ASF Jira:
https://issues.apache.org/jira/browse/SCM-797

**Comment by michael-o** [comment]
@rfscholte Isn't this a solution with can live with on Windows?

**Comment by rfscholte** [comment]
I don't like the idea to commit per file *by default*, this would make the logging unnecessary long and therefore hard to read. If there's no other way to pass these files, we should think of a flag to allow "n files per commit", default "all". I really hope we don't need such flag.

**Comment by murinrad** [comment]
@michael-o indeed this change only changes the "git add" behaviour - it splits it per file so that we don't hit the character limit in the windows CLI. 
A commit is still made only once.

**Comment by rfscholte** [comment]
Kind of surprises me that the GitCheckInCommand contains adders for files, but these seems to be a good way to fix the Windows issue.

**Comment by michael-o** [comment]
@rfscholte Does this mean that you agree with the change?

**Comment by rfscholte** [comment]
Yes, I agree

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/checkin/GitCheckInCommand.java [hunk 1]
```diff
@@ -144,8 +165,11 @@ protected CheckInScmResult executeCheckInCommand( ScmProviderRepository repo, Sc
             {
                 return new CheckInScmResult( null, statusConsumer.getChangedFiles() );
             }
-
-            Commandline clCommit = createCommitCommandLine( repository, fileSet, messageFile );
+            
+            //SCM-714: Workaround for the Windows terminal command limit
+            //
+            Commandline clCommit = createCommitCommandLine( repository, new ScmFileSet( fileSet.getBasedir() ),
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/checkin/GitCheckInCommand.java]: What is the purpose of this change? This affects `-a` to be added, but we did already add untracked changes...

**Comment by michael-o** [comment]
The line in question produces:


    Results :
    
    Failed tests:
      GitExeCheckInCommandTckTest>CheckInCommandTckTest.testCheckInCommandPartialFileset:180     check readme.txt contents expected:<[/readme.txt]> but was:<[changed file]>

Removing the change makes the test pass. Waiting for comment.

**Comment by michael-o** [comment]
@murinrad This is the last open issue for the next release. I'd like to have a comment on the failing test.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/checkin/GitCheckInCommand.java (hunk 1 above)
Comment (murinrad) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/checkin/GitCheckInCommand.java]: The way I read this, the line creates a commit command with files in the fileSet. However sice in the fix I already added all the file I want, I gave it the basedir, which should have no effect. Since the method is public static I didnt want to alter its behaviour so I think thats why I chose this solution. I will check the failing test during the week.

**Comment by murinrad** [comment]
@michael-o seems I accidentaly closed this PR while commiting the fixes.

Indeed the git add was reduntant in createCommitCommandLine and I fixed the empty fileset to its original logic.

The tests which failed previously now pass.

**Comment by michael-o** [comment]
Looks better now. I asking myself what the pupose of the removed else block was.  Commiiting explicitly those files which have been added before makes no sense, because not real subset is selected. In fact, I would have provided the very same long list making it fail again on Windows.

Is that correct?

**Comment by murinrad** [comment]
@michael-o Correct, however with the way I coded the fix previously that method was always receiving an empty filese, hence why it worked. Even without my changes, the else line seems redundant to me.

**Comment by michael-o** [comment]
Thanks for the PR, everything is fine now and merged. I plan to release soon.

**Comment by aharui** [comment]
Is there an ETA for the release?  Are there pre-release builds available we can try?

**Comment by michael-o** [comment]
You can always build from source and change dependency versions.

**Comment by jira-importer** [comment]
Resolve #928
